pageserver/src/tenant/mgr.rs

```diff
@@ -1101,101 +1313,313 @@ pub(crate) async fn attach_tenant(
+            TenantsMap::ShuttingDown(_) => {
+                return Err(TenantMapError::ShuttingDown.into());
             }
-            Err(e) => Err(TenantMapInsertError::Other(e)),
-        },
+            TenantsMap::Open(m) => m,
+        };
+
+        let replaced = m.insert(self.tenant_id, new_value);
+        self.upserted = true;
+
+        // Sanity check: on an upsert we should always be replacing an InProgress marker
+        match replaced {
+            Some(TenantSlot::InProgress(_)) => {
+                // Expected case: we find our InProgress in the map: nothing should have
+                // replaced it because the code that acquires slots will not grant another
+                // one for the same TenantId.
+                Ok(())
+            }
+            None => {
+                error!(
+                    tenant_id = %self.tenant_id,
+                    "Missing InProgress marker during tenant upsert, this is a bug."
+                );
+                Err(TenantSlotUpsertError::InternalError(
+                    "Missing InProgress marker during tenant upsert".into(),
+                ))
+            }
+            Some(slot) => {
+                error!(tenant_id=%self.tenant_id, "Unexpected contents of TenantSlot during upsert, this is a bug.  Contents: {:?}", slot);
+                Err(TenantSlotUpsertError::InternalError(
+                    "Unexpected contents of TenantSlot".into(),
+                ))
+            }
+        }
     }
 }
 
-async fn tenant_map_upsert_slot<'a, F, R>(
-    tenant_id: TenantId,
-    upsert_fn: F,
-) -> Result<(), TenantMapInsertError>
-where
-    F: FnOnce(Option<TenantSlot>) -> R,
-    R: std::future::Future<Output = anyhow::Result<TenantSlot>>,
-{
-    let mut guard = TENANTS.write().await;
-    let m = match &mut *guard {
-        TenantsMap::Initializing => return Err(TenantMapInsertError::StillInitializing),
-        TenantsMap::ShuttingDown(_) => return Err(TenantMapInsertError::ShuttingDown),
+impl Drop for SlotGuard {
+    fn drop(&mut self) {
+        if !self.upserted {
+            let mut locked = TENANTS.write().unwrap();
+
+            let m = match &mut *locked {
+                TenantsMap::Initializing | TenantsMap::ShuttingDown(_) => {
+                    return;
+                }
+                TenantsMap::Open(m) => m,
+            };
+
+            let slot = m.remove(&self.tenant_id);
+            match slot {
+                Some(slot) => match slot {
+                    TenantSlot::InProgress(_) => {
+                        // Normal case: nothing should have replaced the TenantSlot value
+                        // that was set when we were constructed.
+                    }
+                    _ => {
+                        error!(tenant_id=%self.tenant_id, "Unexpected contents of TenantSlot during upsert, this is a bug.  Contents: {:?}", slot);
```

Comment:
Let's follow the same pattern that we had for the deletion queue: have a counter metric for errors that should never happen but we keep going anyway.
(That reminds me: we should configure an alert for that one)

Reply:
This reminds me that we currently had no metrics at all for mgr.rs.  Added a couple to go with the unexpected errors on 2f405efb63783bbdcee6858e1822d62fad676fda